nixos/modules/services/misc/p2pool.nix

```
@@ -0,0 +1,181 @@
+{ config, pkgs, lib, ... }:
+
+let
+  inherit (lib)
+    attrsets
+    concatStringsSep
+    escapeShellArgs
+    literalExpression
+    lists
+    mkDefault
+    mkEnableOption
+    mkIf
+    mkOption
+    mkOptionType
+    length
+    optionalString
+    types;
+
+  inherit (builtins) toString map filter;
+
+  cfg = config.services.p2pool;
+
+  addressesToStringSep = addresses:
+    (concatStringsSep ","
+      (map (x: "${x.address}:${toString x.port}") addresses));
+
+  isListNotEmpty = l: (length l) > 0;
+
+  buildRunCommand = ''
+    ${pkgs.p2pool}/bin/p2pool \
+      ${escapeShellArgs (
+        ["--zmq-port" (toString cfg.zmqPort)]
+        ++ ["--rpc-port" (toString cfg.rpcPort)]
+        ++ ["--host" cfg.host]
+        ++ ["--wallet" cfg.wallet]
+        ++ ["--loglevel" (toString cfg.logLevel)]
+        ++ (lists.optionals (isListNotEmpty cfg.stratumServers) ["--stratum" (addressesToStringSep cfg.stratumServers)])
+        ++ (lists.optionals (isListNotEmpty cfg.p2pServers) ["--p2p" (addressesToStringSep cfg.p2pServers)])
+        ++ (lists.optionals (isListNotEmpty cfg.p2poolNodes) ["--addpeers" (addressesToStringSep cfg.p2poolNodes)])
+        ++ (filter (x: x != "") [
+          (optionalString cfg.lightMode "--light-mode")
+          (optionalString cfg.noCache "--no-cache")
+          (optionalString cfg.noColor "--no-color")
+        ])
+      )
+    }
+  '';
```

Try something like this:
```suggestion
  arguments = escapeShellArgs
    (["--zmq-port" cfg.zmqPort
      "--rpc-port" cfg.rpcPort
      "--host" cfg.host
      "--wallet" cfg.wallet
      "--loglevel" cfg.logLevel
     ] ++ (lists.optionals (isListNotEmpty cfg.stratumServers) ["--stratum" (addressesToStringSep cfg.stratumServers)])
       ++ (lists.optionals (isListNotEmpty cfg.p2pServers) ["--p2p" (addressesToStringSep cfg.p2pServers)])
       ++ (lists.optionals (isListNotEmpty cfg.p2poolNodes) ["--addpeers" (addressesToStringSep cfg.p2poolNodes)])
       ++ (optional cfg.lightMode "--light-mode")
       ++ (optional cfg.noCache "--no-cache")
       ++ (optional cfg.noColor "--no-color"));
```

and do `ExecStart = ${pkgs.p2pool}/bin/p2pool ${arguments}`.